core: add WITHOUT_PEDANTIC() and DECLARE_CONSTANT()

maribu opened the pull request:
### Contribution description

`WITHOUT_PEDANTIC(expr)` disables `-Wpedantic` for `expr`, but switches back to the previous diagnostic settings afterwards. This helps defining macros that are not strictly ISO compliant without having to drop the `-Wpedantic` flag entirely.

`DECLARE_CONSTANT(identifier, const_expr)` declares an anonymous `enum` constant named `identifier` and assigns it the value `const_expr`. Here, `const_expr` has to be a compile time constant, but is not needed to be an integer constant expression. It basically is a tool to magically convert a non-integer constant expression into a integer constant
expression.

### Testing procedure

A unit test was added to check if it works as intended. As this uses `static_assert()`, there is no need to actually run the test.

### Issues/PRs references

Needed by https://github.com/RIOT-OS/RIOT/pull/17177.

benpicco commented on core/include/kernel_defines.h:
```diff
@@ -225,6 +225,33 @@ extern "C" {
+ *
+ * This turns any expression that is constant and known at compile time into
+ * a formal compile time constant. This allows e.g. using non-formally but
+ * still constant expressions in `static_assert()`.
```
This also allows something like

```C
DECLARE_CONSTANT(num_elements, 23);
unsigned foo[num_elements];
```

without creating a VLA right? :tada:

maribu replied:
Yep :-)

benpicco commented:
Can you add a small test to `unittests/tests-kernel_defines` just to make sure this compiles everywhere

maribu commented:
> Can you add a small test to `unittests/tests-kernel_defines` just to make sure this compiles everywhere

Done. Even using `static_assert()` so that the test doesn't need to be run :-)

maribu commented:
For reference:

```
$ cat test.c
#include <stdint.h>

int main(void)
{
    enum { foo = (uintptr_t)((void *) 7) & 3 };
    return 0;
}
$ gcc -std=c11 -Werror -pedantic -o test test.c
test.c: In function 'main':
test.c:5:18: error: enumerator value for 'foo' is not an integer constant expression [-Werror=pedantic]
    5 |     enum { foo = (uintptr_t)((void *) 7) & 3 };
      |                  ^
cc1: all warnings being treated as errors
```

So the constant expression used in the unit tests is not considered an integer constant expression.

maribu commented:
Also: I choose to not temporarily disable `-Wpedantic` in `static_assert()` as this did work with LLVM (but it did work fine with GCC, though). Also, this can be useful to turn variable length arrays into static length arrays or to initialize `const` variables etc.

maribu commented:
I added a comment that the macros are for internal use only and squashed it right in. I don't think that those macros are material for long-time-stable core APIs, so it is better to communicate this.

chrysn reviewed (approved):
Appreciate the warning to keep this internal.

Trusting ben's and the CI tests, ACK.

maribu commented:
Thanks!
